src/TezosProvider.ts

```diff
@@ -0,0 +1,509 @@
+    },
+  ): Promise<SessionTypes.Struct | undefined> => {
+    if (!this.signer || !this.config) {
+      throw new TezosInitializationError();
+    }
+    if (!opts.chains || !opts.chains.length) {
+      throw new TezosProviderError("No chains provided");
+    }
+
+    this.chainId = opts.chains[0].id;
+
+    // convert chain data to map with chain id as a key
+    this.chainMap = opts.chains.reduce((acc, chain) => {
+      acc[chain.id] = chain;
+      return acc;
+    }, {} as ChainsMap);
+
+    let res = await this.signer.connect({
+      namespaces: {
+        tezos: {
+          chains: opts.chains.map((chain) => chain.id),
+          methods: opts.methods ?? DefaultTezosMethods,
+          events: opts.events ?? [],
+        },
+      },
+    });
+    this.isConnected = true;
+
+    const rpcUrl = this.chainMap[this.chainId].rpc[0];
+    this.tezosToolkit = new TezosToolkit(rpcUrl);
+
+    // Set the address if the session exists
+    if (this.signer.session) {
+      let accounts =
+        this.signer.session.namespaces.tezos?.accounts.map((account) => account.split(":")[2]) ??
+        [];
+      if (!accounts.length) {
+        throw new TezosProviderError("No accounts found in session");
+      }
+      // Ensure accounts array is unique
+      this.accounts = [...new Set(accounts)];
+      this.setAddress(this.accounts[0]);
+    }
+    return res;
+  };
+
+  public setAddress = (address: string): void => {
+    if (!this.accounts.includes(address)) {
+      throw new TezosProviderError(
+        `Address ${address} not found in accounts ${this.accounts}. Get Accounts first.`,
+      );
+    }
+    this.address = address;
+  };
+
+  public getChainId = (): string | undefined => {
+    if (!this.config) {
+      throw new TezosInitializationError();
+    }
+    return this.chainId;
+  };
+
+  // Method to get account balance
+  public getBalance = async (): Promise<AssetData> => {
+    if (!this.address) {
+      throw new TezosConnectionError();
+    }
+    if (!this.tezosToolkit) {
+      throw new TezosProviderError("tezosToolkit is not initialized");
+    }
+    const balance = await this.tezosToolkit.tz.getBalance(this.address);
+    const balanceInTez = balance.toNumber();
+    return {
+      balance: balanceInTez,
+      symbol: "ꜩ",
+      name: "XTZ",
+    };
+  };
+
+  public getFormattedBalance = async (): Promise<string> => {
+    const balance = await this.getBalance();
+    return `${balance.balance.toFixed(6)} ꜩ`;
+  };
+
+  public getContractAddress = async (hash: string): Promise<string[]> => {
+    if (!hash) {
+      throw new TezosProviderError(`No hash provided`);
+    }
+
+    const api = this.chainMap[this.chainId].api;
+    const path = `${api}/operations/${hash}`;
+    const response = await axios.get(path);
+    const data = response.data;
+
+    return data
+      .map((op: any) => {
+        const address =
+          op?.status === "applied" && op?.originatedContract?.kind === "smart_contract"
```

Comment:
is it possible that op is undefined | null? this is very complicated and should be cleaned up